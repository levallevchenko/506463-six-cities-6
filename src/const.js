export const SortingTypes = {
  POPULAR: `Popular`,
  PRICE_LOW: `Price: low to high`,
  PRICE_HIGH: `Price: high to low`,
  RATING: `Top rated first`
};

export const CityNames = {
  PARIS: `Paris`,
  COLOGONE: `Cologne`,
  BRUSSELS: `Brussels`,
  AMSTERDAM: `Amsterdam`,
  HAMBURG: `Hamburg`,
  DUSSELDORF: `Dusseldorf`
};

export const Coordinates = {
  [CityNames.PARIS]: {
    latitude: 48.85661,
    longitude: 2.351499,
    zoom: 13
  },
  [CityNames.COLOGNE]: {
    latitude: 50.938361,
    longitude: 6.959974,
    zoom: 13
  },
  [CityNames.BRUSSELS]: {
    latitude: 50.846557,
    longitude: 4.351697,
    zoom: 13
  },
  [CityNames.AMSTERDAM]: {
    latitude: 52.37454,
    longitude: 4.897976,
    zoom: 13
  },
  [CityNames.HAMBURG]: {
    latitude: 53.550341,
    longitude: 10.000654,
    zoom: 13
  },
  [CityNames.DUSSELDORF]: {
    latitude: 51.225402,
    longitude: 6.776314,
    zoom: 13
  }
};

export const CardTypes = {
  MAIN: `MAIN`,
  FAVORITE: `FAVORITE`
};
