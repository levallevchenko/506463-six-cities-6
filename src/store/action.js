export const ActionType = {
  CHANGE_CITY: `offers/changeCity`,
  CHANGE_SORTING: `offers/changeSorting`,
  LOAD_OFFERS: `offers/loadOffers`,
  LOAD_OFFER_DATA: `offers/loadOfferData`,
  LOAD_REVIEWS: `offers/loadReviews`,
  LOAD_NEARBY_OFFERS: `offers/loadNearbyOffers`,
  SET_COMMENT: `offers/setComment`,
  LOAD_COMMENT: `offers/loadComment`,
  SUBMIT_COMMENT: `offers/submitComment`,
  SET_NOT_FOUND_OFFER: `offers/setNotFoundOffer`,
  SET_OFFERS_LOADING_FAILED: `offers/setOffersLoadingFailed`,
  REQUIRED_AUTHORIZATION: `user/requiredAuthorization`,
  SET_AUTH_INFO: `user/login`,
  LOAD_AUTH_INFO: `user/loadAuthInfo`,
  REDIRECT_TO_ROUTE: `app/redirectToRoute`,
  SET_ERROR: `app/setError`,
};

export const ActionCreator = {
  changeCity: (city) => ({
    type: ActionType.CHANGE_CITY,
    payload: city
  }),
  changeSorting: (sorting) => ({
    type: ActionType.CHANGE_SORTING,
    payload: sorting
  }),
  loadOffers: (offers) => ({
    type: ActionType.LOAD_OFFERS,
    payload: offers,
  }),
  loadOfferData: (offer) => ({
    type: ActionType.LOAD_OFFER_DATA,
    payload: offer,
  }),
  loadReviews: (reviews) => ({
    type: ActionType.LOAD_REVIEWS,
    payload: reviews,
  }),
  loadNearbyOffers: (nearby) => ({
    type: ActionType.LOAD_NEARBY_OFFERS,
    payload: nearby,
  }),
  setComment: (comment) => ({
    type: ActionType.SET_COMMENT,
    payload: comment
  }),
  loadComment: (info) => ({
    type: ActionType.LOAD_COMMENT,
    payload: info
  }),
  submitComment: (info) => ({
    type: ActionType.SUBMIT_COMMENT,
    payload: info
  }),
  setNotFoundOffer: () => ({
    type: ActionType.SET_NOT_FOUND_OFFER,
  }),
  setOffersLoadingFailed: () => ({
    type: ActionType.SET_OFFERS_LOADING_FAILED,
  }),
  setAuthInfo: (info) => ({
    type: ActionType.SET_AUTH_INFO,
    payload: info
  }),
  loadAuthInfo: (info) => ({
    type: ActionType.LOAD_AUTH_INFO,
    payload: info
  }),
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: status,
  }),
  redirectToRoute: (url) => ({
    type: ActionType.REDIRECT_TO_ROUTE,
    payload: url
  }),
  setError: (error) => ({
    type: ActionType.SET_ERROR,
    payload: error
  })
};
