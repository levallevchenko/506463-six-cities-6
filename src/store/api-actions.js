import {ActionCreator} from "./action";
import {AuthorizationStatus, APIRoute} from "../const";

export const fetchOffersList = () => (dispatch, _getState, api) => (
  api.get(APIRoute.OFFERS)
    .then(({data}) => dispatch(ActionCreator.loadOffers(data)))
);

export const fetchOffer = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}`)
    .then(({data}) => dispatch(ActionCreator.loadOfferData(data)))
    .catch(() => dispatch(ActionCreator.setNotFoundOffer()))
);

export const fetchReviews = (id) => (dispatch, _getState, api) => (
  api.get(`/comments/${id}`)
    .then(({data}) => dispatch(ActionCreator.loadReviews(data)))
);

export const fetchNearbyOffers = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}/nearby`)
    .then(({data}) => dispatch(ActionCreator.loadNearbyOffers(data)))
);

export const postComment = ({review: comment, rating}, id) => (dispatch, _getState, api) => {
  dispatch(ActionCreator.loadComment(true));
  api.post(`/comments/${id}`, {comment, rating})
    .then(({data}) => dispatch(ActionCreator.setComment(data)))
    .catch(() => {});
};

export const checkAuth = () => (dispatch, _getState, api) => (
  api.get(APIRoute.LOGIN)
    .then(({data}) => dispatch(ActionCreator.setAuthInfo(data)))
    .then(() => dispatch(ActionCreator.loadAuthInfo(true)))
    .then(() => dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH)))
    .catch(() => {})
);

export const login = ({login: email, password}) => (dispatch, _getState, api) => (
  api.post(APIRoute.LOGIN, {email, password})
    .then(({data}) => dispatch(ActionCreator.setAuthInfo(data)))
    .then(() => dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH)))
    .then(() => dispatch(ActionCreator.redirectToRoute(`/`)))
    .catch(() => {})
);
