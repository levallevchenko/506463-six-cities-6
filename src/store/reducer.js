import {CityName, SortingType, AuthorizationStatus} from '../const';
import {ActionType} from './action';
import {adaptOffersData} from '../services/adapter';

const initialState = {
  activeSorting: SortingType.POPULAR,
  activeCity: CityName.PARIS,
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  authInfo: {},
  isOffersLoaded: false,
  offers: [],
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return {
        ...state,
        activeCity: action.payload
      };
    case ActionType.CHANGE_SORTING:
      return {
        ...state,
        activeSorting: action.payload
      };
    case ActionType.LOAD_OFFERS:
      return {
        ...state,
        offers: adaptOffersData(action.payload),
        isOffersLoaded: true
      };
    case ActionType.REQUIRED_AUTHORIZATION:
      return {
        ...state,
        authorizationStatus: action.payload
      };
    case ActionType.SET_AUTH_INFO:
      return {
        ...state,
        authInfo: action.payload
      };
    default:
      return state;
  }
};

export {reducer};
