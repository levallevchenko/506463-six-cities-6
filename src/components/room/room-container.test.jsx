import React from 'react';
import {render} from '@testing-library/react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as redux from 'react-redux';
import {NameSpace} from '../../store/root-reducer';
import {AuthorizationStatus} from '../../const';
import RoomContainer from './room-container';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.spyOn(redux, `useSelector`);
jest.spyOn(redux, `useDispatch`);

it(`RoomContainer should render correctly`, () => {
  const history = createMemoryHistory();

  const store = mockStore({
    [NameSpace.APP]: {
      error: null,
    },
    [NameSpace.USER]: {
      authStatus: AuthorizationStatus.NO_AUTH,
      isAuthInfoLoaded: true,
    },
    [NameSpace.OFFERS]: {
      offer: {
        hotelId: 1,
        location: {latitude: 1, longitude: 1},
        hotelImages: [],
        city: {location: {latitude: 1, longitude: 1}},
        goods: [],
        host: {isUserPro: true}
      },
      reviews: [{user: {avatarUrl: ``}}],
      nearbyOffers: [
        {hotelId: 1, location: {latitude: 1, longitude: 1}},
        {hotelId: 2, location: {latitude: 2, longitude: 2}},
      ],
      offerNotFound: false,
      favoriteOffers: [],
    },
  });

  store.dispatch = () => Promise.resolve();

  const {container} = render(
      <redux.Provider store={store}>
        <Router history={history}>
          <RoomContainer />
        </Router>
      </redux.Provider>
  );

  expect(container).toMatchSnapshot();
});
