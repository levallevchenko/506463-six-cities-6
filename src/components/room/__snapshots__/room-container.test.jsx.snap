// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomContainer should render correctly 1`] = `
<div>
  <div
    class="page"
  >
    <header
      class="header"
      data-testid="header"
    >
      <div
        class="container"
      >
        <div
          class="header__wrapper"
        >
          <div
            class="header__left"
          >
            <a
              class="header__logo-link"
              href="/"
            >
              <img
                alt="6 cities logo"
                class="header__logo"
                data-testid="header-logo"
                height="41"
                src="img/logo.svg"
                width="81"
              />
            </a>
          </div>
          <nav
            class="header__nav"
            data-testid="header-nav"
          >
            <ul
              class="header__nav-list"
            >
              <li
                class="header__nav-item user"
              >
                <a
                  class="header__nav-link header__nav-link--profile"
                  href="/favorites"
                >
                  <div
                    class="header__avatar-wrapper user__avatar-wrapper"
                  />
                  <span
                    class="header__user-name user__name"
                  >
                    Sign in
                  </span>
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
    <main
      class="page__main page__main--property"
    >
      <section
        class="property"
      >
        <div
          class="property__gallery-container container"
        >
          <div
            class="property__gallery"
          />
        </div>
        <div
          class="property__container container"
        >
          <div
            class="property__wrapper"
          >
            <div
              class="property__name-wrapper"
            >
              <h1
                class="property__name"
              />
              <button
                class="property__bookmark-button button"
                type="button"
              >
                <svg
                  class="property__bookmark-icon"
                  height="33"
                  width="31"
                >
                  <use
                    xlink:href="#icon-bookmark"
                  />
                </svg>
                <span
                  class="visually-hidden"
                >
                  To bookmarks
                </span>
              </button>
            </div>
            <div
              class="property__rating rating"
            >
              <div
                class="property__stars rating__stars"
              >
                <span />
                <span
                  class="visually-hidden"
                >
                  Rating
                </span>
              </div>
              <span
                class="property__rating-value rating__value"
              />
            </div>
            <ul
              class="property__features"
            >
              <li
                class="property__feature property__feature--entire"
              />
              <li
                class="property__feature property__feature--bedrooms"
              />
              <li
                class="property__feature property__feature--adults"
              />
            </ul>
            <div
              class="property__price"
            >
              <b
                class="property__price-value"
              >
                €
              </b>
              <span
                class="property__price-text"
              >
                 night
              </span>
            </div>
            <div
              class="property__inside"
            >
              <h2
                class="property__inside-title"
              >
                What's inside
              </h2>
              <ul
                class="property__inside-list"
              />
            </div>
            <div
              class="property__host"
            >
              <h2
                class="property__host-title"
              >
                Meet the host
              </h2>
              <div
                class="property__host-user user"
              >
                <div
                  class="property__avatar-wrapper user__avatar-wrapper property__avatar-wrapper--pro"
                >
                  <img
                    alt="Host avatar"
                    class="property__avatar user__avatar"
                    height="74"
                    width="74"
                  />
                </div>
                <span
                  class="property__user-name"
                />
                <span
                  class="property__user-status"
                >
                  Pro
                </span>
              </div>
              <div
                class="property__description"
              >
                <p
                  class="property__text"
                />
                <p
                  class="property__text"
                >
                  An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.
                </p>
              </div>
            </div>
            <section
              class="property__reviews reviews"
            >
              <h2
                class="reviews__title"
              >
                Reviews · 
                <span
                  class="reviews__amount"
                >
                  1
                </span>
              </h2>
              <ul
                class="reviews__list"
              >
                <li
                  class="reviews__item"
                >
                  <div
                    class="reviews__user user"
                  >
                    <div
                      class="reviews__avatar-wrapper user__avatar-wrapper"
                    >
                      <img
                        alt="Reviews avatar"
                        class="reviews__avatar user__avatar"
                        height="54"
                        src=""
                        width="54"
                      />
                    </div>
                    <span
                      class="reviews__user-name"
                    />
                  </div>
                  <div
                    class="reviews__info"
                  >
                    <div
                      class="reviews__rating rating"
                    >
                      <div
                        class="reviews__stars rating__stars"
                      >
                        <span />
                        <span
                          class="visually-hidden"
                        >
                          Rating
                        </span>
                      </div>
                    </div>
                    <p
                      class="reviews__text"
                    />
                    <time
                      class="reviews__time"
                      datetime=""
                    >
                      May 2021
                    </time>
                  </div>
                </li>
              </ul>
              
            </section>
          </div>
        </div>
        <section
          class="map property__map leaflet-container leaflet-grab leaflet-touch-drag"
          id="map"
          style="height: 579px; position: relative;"
          tabindex="0"
        >
          <div
            class="leaflet-pane leaflet-map-pane"
            style="left: 0px; top: 0px;"
          >
            <div
              class="leaflet-pane leaflet-tile-pane"
            />
            <div
              class="leaflet-pane leaflet-shadow-pane"
            />
            <div
              class="leaflet-pane leaflet-overlay-pane"
            />
            <div
              class="leaflet-pane leaflet-marker-pane"
            />
            <div
              class="leaflet-pane leaflet-tooltip-pane"
            />
            <div
              class="leaflet-pane leaflet-popup-pane"
            />
          </div>
          <div
            class="leaflet-control-container"
          >
            <div
              class="leaflet-top leaflet-left"
            >
              <div
                class="leaflet-control-zoom leaflet-bar leaflet-control"
              >
                <a
                  aria-label="Zoom in"
                  class="leaflet-control-zoom-in"
                  href="#"
                  role="button"
                  title="Zoom in"
                >
                  +
                </a>
                <a
                  aria-label="Zoom out"
                  class="leaflet-control-zoom-out"
                  href="#"
                  role="button"
                  title="Zoom out"
                >
                  −
                </a>
              </div>
            </div>
            <div
              class="leaflet-top leaflet-right"
            />
            <div
              class="leaflet-bottom leaflet-left"
            />
            <div
              class="leaflet-bottom leaflet-right"
            >
              <div
                class="leaflet-control-attribution leaflet-control"
              >
                <a
                  href="https://leafletjs.com"
                  title="A JS library for interactive maps"
                >
                  Leaflet
                </a>
              </div>
            </div>
          </div>
        </section>
      </section>
      <div
        class="container"
      >
        <section
          class="near-places places"
        >
          <h2
            class="near-places__title"
          >
            Other places in the neighbourhood
          </h2>
          <div
            class="near-places__list places__list"
          >
            <article
              class="place-card cities__place-card"
            >
              <div
                class="place-card__image-wrapper cities__image-wrapper"
              >
                <a
                  href="/offer/1"
                >
                  <img
                    alt="Place image"
                    class="place-card__image"
                    height="200"
                    width="260"
                  />
                </a>
              </div>
              <div
                class="place-card__info"
              >
                <div
                  class="place-card__price-wrapper"
                >
                  <div
                    class="place-card__price"
                  >
                    <b
                      class="place-card__price-value"
                    >
                      €
                    </b>
                    <span
                      class="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    class="place-card__bookmark-button button"
                    type="button"
                  >
                    <svg
                      class="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlink:href="#icon-bookmark"
                      />
                    </svg>
                    <span
                      class="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  class="place-card__rating rating"
                >
                  <div
                    class="place-card__stars rating__stars"
                  >
                    <span />
                    <span
                      class="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  class="place-card__name"
                >
                  <a
                    href="/offer/1"
                  />
                </h2>
                <p
                  class="place-card__type"
                />
              </div>
            </article>
            <article
              class="place-card cities__place-card"
            >
              <div
                class="place-card__image-wrapper cities__image-wrapper"
              >
                <a
                  href="/offer/2"
                >
                  <img
                    alt="Place image"
                    class="place-card__image"
                    height="200"
                    width="260"
                  />
                </a>
              </div>
              <div
                class="place-card__info"
              >
                <div
                  class="place-card__price-wrapper"
                >
                  <div
                    class="place-card__price"
                  >
                    <b
                      class="place-card__price-value"
                    >
                      €
                    </b>
                    <span
                      class="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    class="place-card__bookmark-button button"
                    type="button"
                  >
                    <svg
                      class="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlink:href="#icon-bookmark"
                      />
                    </svg>
                    <span
                      class="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  class="place-card__rating rating"
                >
                  <div
                    class="place-card__stars rating__stars"
                  >
                    <span />
                    <span
                      class="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  class="place-card__name"
                >
                  <a
                    href="/offer/2"
                  />
                </h2>
                <p
                  class="place-card__type"
                />
              </div>
            </article>
          </div>
        </section>
      </div>
    </main>
  </div>
</div>
`;
